File: laundry-reservation/typescript/src/MachineApi.ts
import type { MachineDevice } from './MachineDevice.js';

export class MachineApi {
  private readonly devices = new Map<number, MachineDevice>();
  private readonly reservationToMachine = new Map<string, number>();

  registerDevice(machineNumber: number, device: MachineDevice): void {
    this.devices.set(machineNumber, device);
  }

  lock(reservationId: string, machineNumber: number, reservationDateTime: Date, pin: number): boolean {
    const device = this.devices.get(machineNumber);
    if (!device) return false;

    if (this.reservationToMachine.has(reservationId)) {
      device.lock(reservationId, reservationDateTime, pin);
      return true;
    }

    const locked = device.lock(reservationId, reservationDateTime, pin);
    if (locked) {
      this.reservationToMachine.set(reservationId, machineNumber);
    }
    return locked;
  }

  unlock(machineNumber: number, reservationId: string): void {
    const device = this.devices.get(machineNumber);
    if (device) {
      device.unlock(reservationId);
      this.reservationToMachine.delete(reservationId);
    }
  }
}

File: laundry-reservation/csharp/src/LaundryReservation/MachineApi.cs
namespace LaundryReservation;

public class MachineApi
{
    private readonly Dictionary<int, IMachineDevice> _devices = new();
    private readonly Dictionary<string, int> _reservationToMachine = new();

    public void RegisterDevice(int machineNumber, IMachineDevice device)
    {
        _devices[machineNumber] = device;
    }

    public bool Lock(string reservationId, int machineNumber, DateTime reservationDateTime, int pin)
    {
        if (!_devices.TryGetValue(machineNumber, out var device))
            return false;

        if (_reservationToMachine.ContainsKey(reservationId))
        {
            device.Lock(reservationId, reservationDateTime, pin);
            return true;
        }

        var locked = device.Lock(reservationId, reservationDateTime, pin);
        if (locked)
        {
            _reservationToMachine[reservationId] = machineNumber;
        }
        return locked;
    }

    public void Unlock(int machineNumber, string reservationId)
    {
        if (_devices.TryGetValue(machineNumber, out var device))
        {
            device.Unlock(reservationId);
            _reservationToMachine.Remove(reservationId);
        }
    }
}
